ipn/store.go

```diff
@@ -14,14 +14,73 @@ import (
 	"log"
 	"os"
 	"path/filepath"
+	"strings"
 	"sync"
 	"time"
 
 	"tailscale.com/atomicfile"
 	"tailscale.com/kube"
+	_ "tailscale.com/kube"
 	"tailscale.com/paths"
 )
 
+// NewStoreFunc returns a StateStore for the provided path.
+type NewStoreFunc func(path string) (StateStore, error)
+
+var knownStores = map[string]NewStoreFunc{
+	"mem:": func(string) (StateStore, error) {
+		return new(MemoryStore), nil
+	},
+	"kube:": func(path string) (StateStore, error) {
+		secretName := strings.TrimPrefix(path, "kube:")
+		return NewKubeStore(secretName)
+	},
+	"file:": func(path string) (StateStore, error) {
+		path = strings.TrimPrefix(path, "file:")
+		return NewFileStore(path)
+	},
+}
+
+// ErrNoMatchingStore is the error used when there is no matching
+// registered store.
+var ErrNoMatchingStore = errors.New("no matching store")
+
+// NewStore returns a StateStore based on the provided path
+// and registered stores.
+func NewStore(path string) (StateStore, error) {
+	for prefix, sf := range knownStores {
+		if strings.HasPrefix(path, prefix) {
+			// We can't strip the prefix here as some NewStoreFunc (like arn:)
+			// expect the prefix.
+			return sf(path)
+		}
+	}
+	return nil, fmt.Errorf("%v: %w", path, ErrNoMatchingStore)
+}
+
+// RegisterStore registers a prefix to be used for
+// NewStore. It panics if the prefix is empty, or has
+// already been defined.
+// The provided fn is called with the path passed to NewStore,
+// the prefix is not stripped.
+// By default the following stores are registered:
+//
+//   * if the string begins with "mem:", the suffix
+//     is ignored and an in-memory store is used.
+//   * if the string begins with "kube:", the suffix
```

Comment:
This isn't "by default". In the other file in this PR you had to blank import it to get the side effect of its registration?

Reply:
These are the default ones (are in the same package), I had to blank import the aws one.